tss-esapi/src/abstraction/public.rs

```diff
@@ -3,126 +3,232 @@
+    AffinePoint,
+    CurveArithmetic,
+    FieldBytesSize,
+    PublicKey,
 };
+use rsa::{pkcs8::EncodePublicKey, BigUint, RsaPublicKey};
+use x509_cert::spki::SubjectPublicKeyInfoOwned;
 
-/// Can be converted from [`crate::structures::Public`] when not a fully constructed
-/// [`picky_asn1_x509::SubjectPublicKeyInfo`] is required.
-///
-/// # Details
-/// Holds either [`picky_asn1_x509::RsaPublicKey`] for [`crate::structures::Public::Rsa`] or
-/// [`picky_asn1_x509::EcPoint`] for [`crate::structures::Public::Ecc`].
-///
-/// This object can be serialized and deserialized
-/// using serde if the `serde` feature is enabled.
-#[derive(Debug, PartialEq, Eq, Clone)]
-#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
-pub enum DecodedKey {
-    RsaPublicKey(RsaPublicKey),
-    EcPoint(EcPoint),
+impl<C> TryFrom<&Public> for PublicKey<C>
+where
+    C: CurveArithmetic + AssociatedTpmCurve,
+    FieldBytesSize<C>: ModulusSize,
+    AffinePoint<C>: FromEncodedPoint<C> + ToEncodedPoint<C>,
+{
+    type Error = Error;
+
+    fn try_from(value: &Public) -> Result<Self, Self::Error> {
+        match value {
+            Public::Ecc {
+                parameters, unique, ..
+            } => {
+                if parameters.ecc_curve() != C::TPM_CURVE {
+                    return Err(Error::local_error(WrapperErrorKind::InvalidParam));
+                }
+
+                let x = unique.x().as_bytes();
+                let y = unique.y().as_bytes();
+
+                // TODO: When elliptic_curve bumps to 0.14, we can use the TryFrom implementation instead
+                // of checking lengths manually
+                if x.len() != FieldBytesSize::<C>::USIZE {
+                    return Err(Error::local_error(WrapperErrorKind::InvalidParam));
+                }
+                if y.len() != FieldBytesSize::<C>::USIZE {
+                    return Err(Error::local_error(WrapperErrorKind::InvalidParam));
+                }
+
+                let encoded_point =
+                    EncodedPoint::<C>::from_affine_coordinates(x.into(), y.into(), false);
+                let public_key = PublicKey::<C>::try_from(&encoded_point)
+                    .map_err(|_| Error::local_error(WrapperErrorKind::InvalidParam))?;
+
+                Ok(public_key)
+            }
+            _ => Err(Error::local_error(WrapperErrorKind::UnsupportedParam)),
+        }
+    }
 }
 
-impl TryFrom<Public> for DecodedKey {
+impl TryFrom<&Public> for RsaPublicKey {
     type Error = Error;
 
-    fn try_from(value: Public) -> Result<Self, Self::Error> {
-        public_to_decoded_key(&value)
+    fn try_from(value: &Public) -> Result<Self, Self::Error> {
+        match value {
+            Public::Rsa {
+                unique, parameters, ..
+            } => {
+                let exponent = match parameters.exponent() {
+                    RsaExponent::ZERO_EXPONENT => BigUint::from(65537u32),
```

Comment:
Can you make this into a constant instead with a name that makes it more clear what it is. Because it does not seem to be ```u16::MAX```.